File: src/main/java/org/gecko/viewmodel/EdgeViewModel.java
package org.gecko.viewmodel;

import javafx.beans.binding.Bindings;
import javafx.beans.property.*;
import javafx.geometry.Point2D;
import lombok.Getter;
import lombok.NonNull;
import lombok.Setter;
import org.gecko.exceptions.ModelException;
import org.gecko.model.Edge;
import org.gecko.model.Kind;

/**
 * Represents an abstraction of an {@link Edge} model element. An {@link EdgeViewModel} is described by a source- and a
 * destination-{@link StateViewModel}. It is also associated with one of the start-{@link StateViewModel}'s
 * {@link ContractViewModel}s, has a priority and a {@link Kind}, which informs about how the associated
 * {@link ContractViewModel} is handled. Contains methods for managing the afferent data and updating the
 * target-{@link Edge}.
 */
@Getter
@Setter
public class EdgeViewModel extends PositionableViewModelElement<Edge> {

    private final ObjectProperty<Kind> kindProperty = new SimpleObjectProperty<>();
    private final IntegerProperty priorityProperty = new SimpleIntegerProperty();
    private final ObjectProperty<ContractViewModel> contractProperty = new SimpleObjectProperty<>();
    private final ObjectProperty<StateViewModel> sourceProperty = new SimpleObjectProperty<>();
    private final ObjectProperty<StateViewModel> destinationProperty = new SimpleObjectProperty<>();
    private final BooleanProperty isLoopProperty = new SimpleBooleanProperty();
    private final IntegerProperty orientationProperty = new SimpleIntegerProperty();
    /**
     * The list of edge points that define the path of the edge.
     */

    private final ObjectProperty<Point2D> startPointProperty = new SimpleObjectProperty<>();
    private final ObjectProperty<Point2D> endPointProperty = new SimpleObjectProperty<>();
    private final ObjectProperty<Point2D> startOffsetProperty = new SimpleObjectProperty<>(Point2D.ZERO);
    private final ObjectProperty<Point2D> endOffsetProperty = new SimpleObjectProperty<>(Point2D.ZERO);

    public EdgeViewModel(int id, @NonNull Edge target, @NonNull StateViewModel source, @NonNull StateViewModel destination) {
        super(id, target);
        setKind(target.getKind());
        setPriority(target.getPriority());
        sourceProperty.set(source);
        destinationProperty.set(destination);
        isLoopProperty.bind(Bindings.createBooleanBinding(() -> getSource().equals(getDestination()), sourceProperty,
                destinationProperty));

        sizeProperty.setValue(Point2D.ZERO);
        setBindings();

        getSource().getOutgoingEdges().add(this);
        getDestination().getIncomingEdges().add(this);
    }

    public void setBindings() {
        startPointProperty.bind(Bindings.createObjectBinding(
                () -> getSource().getCenter().add(startOffsetProperty.getValue()), startOffsetProperty, getSource().getPositionProperty()));
        endPointProperty.bind(Bindings.createObjectBinding(
                () -> getDestination().getCenter().add(endOffsetProperty.getValue()), endOffsetProperty, getDestination().getPositionProperty()));
    }

    private void removeBindings() {
        startPointProperty.unbind();
        endPointProperty.unbind();
    }

    public void setStartOffsetProperty(Point2D startOffset) {
        this.startOffsetProperty.setValue(startOffset);
    }

    public void setEndOffsetProperty(Point2D endOffset) {
        this.endOffsetProperty.setValue(endOffset);
    }

    public void setPriority(int priority) {
        priorityProperty.setValue(priority);
    }

    public int getPriority() {
        return priorityProperty.getValue();
    }

    public void setKind(@NonNull Kind kind) {
        kindProperty.setValue(kind);
    }

    public Kind getKind() {
        return kindProperty.getValue();
    }

    public void setContract(ContractViewModel contract) {
        contractProperty.setValue(contract);
    }

    public ContractViewModel getContract() {
        return contractProperty.getValue();
    }

    public void setSource(@NonNull StateViewModel source) {
        getSource().getOutgoingEdges().remove(this);
        removeBindings();
        sourceProperty.setValue(source);
        setBindings();
        source.getOutgoingEdges().add(this);
    }

    public StateViewModel getSource() {
        return sourceProperty.getValue();
    }

    public void setDestination(@NonNull StateViewModel destination) {
        getDestination().getIncomingEdges().remove(this);
        removeBindings();
        destinationProperty.setValue(destination);
        setBindings();
        destination.getIncomingEdges().add(this);
    }

    public StateViewModel getDestination() {
        return destinationProperty.getValue();
    }

    @Override
    public void updateTarget() throws ModelException {
        target.setKind(getKind());
        target.setPriority(getPriority());
        if (contractProperty.getValue() != null) {
            target.setContract(contractProperty.getValue().getTarget());
        }
        target.setSource(sourceProperty.getValue().getTarget());
        target.setDestination(destinationProperty.getValue().getTarget());
    }

    @Override
    public <S> S accept(@NonNull PositionableViewModelElementVisitor<S> visitor) {
        return visitor.visit(this);
    }

    /**
     * Returns a string representation of this {@link EdgeViewModel} in the form of "priority. kind(contract)".
     *
     * @return a string representation of this {@link EdgeViewModel}
     */
    public String getRepresentation() {
        String representation = "";
        representation += getPriority() + ". ";
        representation += getKind().name();
        if (getContract() != null) {
            representation += "(" + getContract().getName() + ")";
        }
        return representation;
    }

    public Point2D getStartPoint() {
        return startPointProperty.getValue();
    }

    public Point2D getEndPoint() {
        return endPointProperty.getValue();
    }

    public void setStartPoint(Point2D startPoint) {
        removeBindings();
        startPointProperty.setValue(startPoint);
    }

    public void setEndPoint(Point2D endPoint) {
        removeBindings();
        endPointProperty.setValue(endPoint);
    }

    public boolean isLoop() {
        return isLoopProperty.getValue();
    }

    public void setOrientation(int orientation) {
        orientationProperty.setValue(orientation);
    }

    @Override
    public void setPosition(@NonNull Point2D position) {
    }

    @Override
    public void setSize(@NonNull Point2D position) {
    }

    @Override
    public Point2D getCenter() {
        return startPointProperty.getValue().midpoint(endPointProperty.getValue());
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (!(o instanceof EdgeViewModel edge)) {
            return false;
        }
        return id == edge.id;
    }
}

File: src/main/kotlin/org/gecko/viewmodel/EdgeViewModel.kt
package org.gecko.viewmodel


import com.google.gson.JsonNull
import com.google.gson.JsonPrimitive
import javafx.beans.binding.Bindings
import javafx.beans.property.*
import javafx.beans.value.ObservableValue
import javafx.geometry.Point2D
import org.gecko.actions.ActionManager
import org.gecko.view.GeckoView
import org.gecko.view.contextmenu.EdgeViewElementContextMenuBuilder
import org.gecko.view.inspector.builder.EdgeInspectorBuilder
import org.gecko.view.views.viewelement.EdgeViewElement
import org.gecko.view.views.viewelement.decorator.ConnectionElementScalerViewElementDecorator
import org.gecko.view.views.viewelement.decorator.ViewElementDecorator
import tornadofx.getValue
import tornadofx.setValue

/**
 * Represents an abstraction of an [Edge] model element. An [EdgeViewModel] is described by a source- and a
 * destination-[StateViewModel]. It is also associated with one of the start-[StateViewModel]'s
 * [ContractViewModel]s, has a priority and a [Kind], which informs about how the associated
 * [ContractViewModel] is handled. Contains methods for managing the afferent data and updating the
 * target-[Edge].
 */
class EdgeViewModel(src: StateViewModel, dst: StateViewModel) : PositionableViewModelElement(), Inspectable {
    val kindProperty: ObjectProperty<Kind> = SimpleObjectProperty(Kind.HIT)
    val priorityProperty: IntegerProperty = SimpleIntegerProperty(0)
    val contractProperty: ObjectProperty<ContractViewModel?> = nullableObjectProperty()
    val sourceProperty: ObjectProperty<StateViewModel> = objectProperty(src)
    val destinationProperty: ObjectProperty<StateViewModel> = objectProperty(dst)
    val isLoopProperty: BooleanProperty = SimpleBooleanProperty(false)
    val orientationProperty: IntegerProperty = SimpleIntegerProperty(0)

    var priority: Int by priorityProperty
    var kind: Kind by kindProperty
    var contract: ContractViewModel? by contractProperty
    var source: StateViewModel by sourceProperty
    var destination: StateViewModel by destinationProperty

    override fun asJson() = super.asJson().apply {
        addProperty("source", source.name)
        addProperty("destination", destination.name)
        add("contract",
            contract?.name?.let { JsonPrimitive(it) } ?: JsonNull.INSTANCE)
        addProperty("kind", kind.name)
        addProperty("priority", priority)
    }


    /**
     * The list of edge points that define the path of the edge.
     */
    val startOffsetProperty: ObjectProperty<Point2D> = SimpleObjectProperty(Point2D.ZERO)
    val startPointProperty: ObservableValue<Point2D> = Bindings.createObjectBinding(
        { source.center.add(startOffsetProperty.value) },
        startOffsetProperty,
        source.positionProperty
    )
    val endOffsetProperty: ObjectProperty<Point2D> = SimpleObjectProperty(Point2D.ZERO)
    val endPointProperty: ObservableValue<Point2D> = Bindings.createObjectBinding(
        { destination.center.add(endOffsetProperty.value) },
        endOffsetProperty,
        destinationProperty,
        destination.positionProperty
    )


    val representation: String
        /**
         * Returns a string representation of this [EdgeViewModel] in the form of "priority. kind(contract)".
         *
         * @return a string representation of this [EdgeViewModel]
         */
        get() {
            var representation = ""
            representation += priority.toString() + ". "
            representation += kind.name
            if (contract != null) {
                representation += "(" + contract!!.name + ")"
            }
            return representation
        }

    val startPoint: Point2D by startPointProperty
    val endPoint: Point2D by endPointProperty

    val isLoop: Boolean
        get() = isLoopProperty.value

    init {
        startPointProperty

        endPointProperty


        sourceProperty.onChange { field, new ->
            field?.outgoingEdges!!.remove(this)
            removeBindings()
            setBindings()
            new?.outgoingEdges?.add(this)
        }

        destinationProperty.onChange { field, new ->
            field?.incomingEdges?.remove(this)
            removeBindings()
            setBindings()
            new?.incomingEdges?.add(this)
        }

        sourceProperty.set(source)
        destinationProperty.set(destination)
        isLoopProperty.bind(
            Bindings.createBooleanBinding(
                { this.source == this.destination }, sourceProperty,
                destinationProperty
            )
        )

        sizeProperty.value = Point2D.ZERO
        setBindings()

        this.source.outgoingEdges.add(this)
        this.destination.incomingEdges?.add(this)
    }

    fun setBindings() {}

    fun removeBindings() {
        //startPointProperty.unbind()
        //endPointProperty.unbind()
    }

    fun setStartOffsetProperty(startOffset: Point2D) {
        startOffsetProperty.value = startOffset
    }

    fun setEndOffsetProperty(endOffset: Point2D) {
        endOffsetProperty.value = endOffset
    }

    override val children: Sequence<Element>
        get() = sequenceOf()

    override fun view(actionManager: ActionManager, geckoView: GeckoView): ViewElementDecorator {
        val newEdgeViewElement = EdgeViewElement(this)
        val contextMenuBuilder = EdgeViewElementContextMenuBuilder(actionManager, this, geckoView)
        //setContextMenu(newEdgeViewElement.pane, contextMenuBuilder)
        return ConnectionElementScalerViewElementDecorator(newEdgeViewElement)
    }

    fun setOrientation(orientation: Int) {
        orientationProperty.value = orientation
    }

    override val center: Point2D
        get() = startPointProperty.value.midpoint(endPointProperty.value)

    override fun inspector(actionManager: ActionManager) = EdgeInspectorBuilder(actionManager, this)
}
